src/Data/AudienceTree.elm

```diff
@@ -0,0 +1,361 @@
+      }
+    )
+
+
+
+-- STEP 1
+--
+-- Transform the list of `AudienceFolder` into a dictionnary `Id -> PendingFolder`
+-- each `PendingFolder` being "empty": only name and parent are provided.
+--
+-- It would be problematic if multiple folders share the same id `I`. If an audience
+-- belongs to `I`, to what folder should we put it? If we encounter
+-- such a case, only the first folder is kept and we keep track of those problematic folders
+-- in the `duplicatedFolderIds` list.
+
+
+type alias Step1Result =
+    { root : PendingRootFolder
+    , otherFolders : Dict Int PendingFolder
+    , duplicatedFolderIds : List DuplicatedIdFolder
+    }
+
+
+
+-- A PendingFolder is a temporary value used to build the final tree.
+-- It will accumulate all the audiences belonging to it (during the "STEP 2")
+
+
+type alias PendingFolder =
+    { id : Int
+    , name : String
+    , audiences : List String
+    , parent : Maybe Int
+    }
+
+
+type alias DuplicatedIdFolder =
+    { name : String
+    , duplicates : PendingFolder
+    }
+
+
+type alias PendingRootFolder =
+    { name : String
+    , audiences : List String
+    }
+
+
+step1 : List AudienceFolder -> Step1Result
+step1 folders =
+    List.foldl
+        (\folder ( others, duplicated ) ->
+            case Dict.get folder.id others of
+                Nothing ->
+                    ( Dict.insert folder.id
+                        { id = folder.id
+                        , name = folder.name
+                        , audiences = []
+                        , parent = folder.parent
+                        }
+                        others
+                    , duplicated
+                    )
+
+                Just pendingFolder ->
+                    ( others
+                    , { name = folder.name
+                      , duplicates = pendingFolder
+                      }
+                        :: duplicated
+                    )
+        )
+        ( Dict.empty, [] )
+        folders
+        |> (\( others, duplicated ) ->
+                { root =
+                    { name = "Root"
+                    , audiences = []
+                    }
+                , otherFolders = others
+                , duplicatedFolderIds = duplicated
+                }
+           )
+
+
+
+-- STEP 2
+--
+-- In this step, we add all the audiences to their corresponding PendingFolder.
+-- If the folder indicated in the audience record doesn't exist, the audience
+-- is assigned to root and we keep track of this audience in the
+-- `missingFolderId` list.
+--
+-- This step has to be performed AFTER the previous one (we have to know all the
+-- folders ids). This is "softly constrained" by the fact that the `step2`
+-- function takes `Step1Result` as argument (we might use some opaque type technique
+-- to have a "hard constaint" about function call order but this would split the process
+-- across multiple files which doesn't seem to worth it for such a small project).
```

Comment:
> which doesn't seem worth it for such a small project

Honestly, even your three-step process with the error handling is quite a bit of overkill IMHO :joy: In normal code I'd perhaps consider relying on BE guarantees more (no duplicities, no folders/audiences with parent that doesn't exist) but it's cool that you're showing this off!